### DISCUSSION TITLE
How to enable multiple file extensions in storage policy

### DISCUSSION THREAD

**Discussion by mationai** [supabase/supabase#1787] [Questions] [answered]
(storage-js issues redirected me here)

In `Adding new policy to images` screen, I see this as the first line:

`storage."extension"(name) = 'jpg'`

Is there a way to change this to accept all image file extensions?

**Comment by inian** [comment, marked as answer]
Hi @fuzzthink, 

You can whitelist multiple file extensions like this
```
storage."extension"(name) in ('jpg', 'png', 'webp')
```

I don't think there is a way to whitelist all image files other than this, but there aren't that many image formats that the browser / mobile apps can render, so hopefully this is fine.

**Reply by mationai** [reply]
Thank you. I did that and now it errors with statusCode "42501" - 'new row violates row-level security policy for table "objects"'.

I searched passed discussions and issues and they don't seem to be storage related. I don't see anything in docs mentioning setting policy for "objects" table either.

**Reply by inian** [reply]
Can you show the full policy you have and the row you are trying to insert John?

**Reply by mationai** [reply]
Policy on "images" bucket:
```sql
(storage.extension(name) = ANY (ARRAY['jpg'::text, 'jpeg'::text, 'svg'::text, 'png'::text, 'webp'::text, 'gif'::text])) AND (lower((storage.foldername(name))[1]) = 'public'::text) AND (role() = 'anon'::text)
```

code:
```js
await supabase.storage.from('images').upload('public/photo.svg', file)
```

It POST to `...supabase.co/storage/v1/object/images/public/photo.svg`

I don't have any policy on "object", unless it's hidden somewhere, I don't think the app UI gives me the ability.

Using the latest, v1.12.0

**Reply by inian** [reply]
This policy needs to be added to the objects table. No policy required for the buckets table.

**Reply by mationai** [reply]
Can you please show me the policy statement to apply on objects table so I can write to the images bucket?

**Reply by inian** [reply]
The policy editor from the UI by default attaches the policy to the objects table. So on the storage policy editor page, click on create a policy from scratch, paste your policy definition
```
(storage.extension(name) = ANY (ARRAY['jpg', 'jpeg', 'svg', 'png', 'webp', 'gif'])) AND (lower((storage.foldername(name))[1]) = 'public') AND (role() = 'anon')
```
Click on `insert` in the allowed operations and when you click on review you should see something like this
![image](https://user-images.githubusercontent.com/2155545/120429864-b562b880-c3a8-11eb-843e-89cafeb5ad82.png)

**Reply by mationai** [reply]
If I don't want 'anon' users to insert, shouldn't `insert` left unchecked? The client code is made with the auth user, so shouldn't it be successful w/o adding `insert` policy for 'anon' users?

**Reply by inian** [reply]
Even auth users need the right policies to be able to upload objects. Unless you are using the serviceKey which bypasses RLS, all other users need the right permissions. 
Looks like you need 2 policies, correct me if i am wrong
- all users (anon and auth) to read images from public folder
```
-- SELECT permission
(storage.extension(name) = ANY (ARRAY['jpg', 'jpeg', 'svg', 'png', 'webp', 'gif'])) AND (lower((storage.foldername(name))[1]) = 'public')
```
- auth users to upload images to public folder
```
-- INSERT permission
(storage.extension(name) = ANY (ARRAY['jpg', 'jpeg', 'svg', 'png', 'webp', 'gif'])) AND (lower((storage.foldername(name))[1]) = 'public') AND (role() = 'authenticated')
```

**Reply by mationai** [reply]
Thank you @inian, it works. For reference, the comments need to be removed, as it will comment out the code in preview window.
